01-messaging.md

```diff
@@ -387,6 +394,234 @@ every message maximally).
 Finally, the usage of periodic `ping` messages serves to promote frequent key
 rotations as specified within [BOLT #8](08-transport.md).
 
+## Appendix A: Type-Length-Value Test Vectors
+
+The following tests assume that two separate TLV namespaces exist: n1 and n2.
+
+The n1 namespace supports the following TLV types:
+
+1. tlvs: `n1`
+2. types:
+   1. type: 1 (`tlv1`)
+   2. data:
+     * [`tu64`:`amount_msat`]
+   1. type: 2 (`tlv2`)
+   2. data:
+     * [`short_channel_id`:`scid`]
+   1. type: 3 (`tlv3`)
+   2. data:
+     * [`point`:`node_id`]
+     * [`u64`:`amount_msat_1`]
+     * [`u64`:`amount_msat_2`]
+   1. type: 14 (`tlv4`)
+   2. data:
+     * [`u16`:`cltv_delta`]
+
+The n2 namespace supports the following TLV types:
+
+1. tlvs: `n2`
+2. types:
+   1. type: 10 (`tlv1`)
+   2. data:
+     * [`tu64`:`amount_msat`]
+   1. type: 11 (`tlv2`)
+   2. data:
+     * [`tu32`:`cltv_expiry`]
+
+### TLV Decoding Failures
+
+The following TLV streams in any namespace should trigger a decoding failure:
+
+1. Invalid stream: 0xfd
+2. Reason: type truncated
+
+1. Invalid stream: 0xfd01
+2. Reason: type truncated
+
+1. Invalid stream: 0xfd0001 01
```

Comment:
```suggestion
1. Invalid stream: 0xfd0100
```
the current value is minimally encoded, remove trailing 0x01

Reply:
> Thanks for getting the ball rolling on the test vectors @rustyrussell! I'm about half way through, but the main thing i'm noticing so far is a discrepancy in the way that the varints are being encoded. `CompactSize` encodes any multi-byte values using little endian, while the tests are using big-endian.

You're right :(  Bitcoin strikes again...

I assume we should follow (and document!) Bitcoin CompactSize endian here, since that was the rationale for not inventing our own?

Reply:
It should be implicit from the test vectors, but if we want to be more specific fine by me :)

fwiw I have a series of varint test vectors, should I make a separate pr with those?

Reply:
Good catch @cfromknecht !
That shows I went too quickly when comparing my test suite and this PR, I'll spend more time on it today.

Reply:
Unfortunately, my tests *passed* after I fixed the endian without fixing the data.  I'll add some test vectors which *only* work in the correct endian.

Reply:
And we need the trailing 01, otherwise it is invalid for a different reason: no length field.  Though that should be 00 not 01.

Reply:
shouldn't the check for whether the type is minimally encoded be applied before attempting to parse the length?